docs/source/tutorials/chat.rst

```diff
@@ -0,0 +1,164 @@
+(BOS) and end-of-sequence (EOS) tokens that are represented differently in the tokenizer
+than the rest of the prompt template. Let's tokenize some examples with the
+:class:`~torchtune.modules.tokenizers.SentencePieceTokenizer` used by LLaMA2 to see
+why.
+
+.. code-block:: python
+
+    from torchtune.modules.tokenizers import SentencePieceTokenizer
+
+    tokenizer = SentencePieceTokenizer("/tmp/Llama-2-7b-hf/tokenizer.model")
+    text = "[INST] <<SYS>> Hello"
+    tokens = tokenizer.encode(text, add_bos=True, add_eos=True)
+    print(tokens)
+    # [1, 518, 25580, 29962, 3532, 14816, 29903, 6778, 15043, 2]
+
+We've added the BOS and EOS tokens when encoding our example text. This shows up
+as IDs 1 and 2. We can verify that these are our BOS and EOS tokens.
+
+.. code-block:: python
+
+    print(tokenizer.spm_model.piece_to_id("<s>"))
+    # 1
+    print(tokenizer.spm_model.piece_to_id("</s>"))
+    # 2
+
+The BOS and EOS tokens are what we call special tokens, because they have their own
+reserved token IDs. This means that they will index to their own individual vectors in
+the model's learnt embedding table. The rest of the prompt template tags, :code:`[INST]`
+and :code:`<<SYS>>` are tokenized as normal text and not their own IDs.
+
+.. code-block:: python
+
+    print(tokenizer.decode(518))
+    # '['
+    print(tokenizer.decode(25580))
+    # 'INST'
+    print(tokenizer.decode(29962))
+    # ']'
+
+It's important to note that you should not place the special reserved tokens in your
+input prompts manually, as it will be treated as normal text and not as a special
+token.
+
+.. code-block:: python
+
+    print(tokenizer.encode("<s>", add_bos=False, add_eos=False))
+    # [529, 29879, 29958]
+
+When should I use a prompt template?
+------------------------------------
+
+When to use or not to use a prompt template is governed by what your desired inference
+behavior is. You should use a prompt template in the following scenarios:
+
+1. You are running inference on the base model and it was pre-trained with a prompt
+template
+2. You want to prime a fine-tuned model to expect a certain prompt structure on inference
+for a specific task
+
+It is not strictly necessary to fine-tune with a prompt template, but generally you
+want the model to perform some sort of task, which will require some formatting of
+the prompt.
+
+For example, the :class:`~torchtune.data.SummarizeTemplate` provides a lightweight
+structure to prime your fine-tuned model for prompts asking to summarize text.
+
+.. code-block:: python
+
+    f"Summarize this dialogue:\n{dialogue}\n---\nSummary:\n"
+
+You can fine-tune LLaMA2 with this template even though the model was originally pre-trained
+with the :class:`~torchtune.data.Llama2ChatFormat`, as long as this is what the model
+sees during inference. The model should be robust enough to adapt to a new template.
+
+Special tokens in LLaMA3
```

Comment:
These sections don't flow together super clearly. Prompt template + special tokens section, followed by when to use prompt template, followed by specific handling of special tokens for Llama3. Why not just tackle prompt template and special tokens as their own standalone entities and walk through them fully one at a time?

Reply:
Yeah, let me see if I can separate these two a bit better.